@lindsayad I can't figure out my compile problem, can you take a look?

```
undefined reference to `RankThreeTensorTempl<double>::RankThreeTensorTempl()'
undefined reference to `RankThreeTensorTempl<double>::doubleContraction(RankTwoTensorTempl<double> const&) const'
```

--- a/framework/include/restart/DataIO.h
+++ b/framework/include/restart/DataIO.h
@@ -17,6 +17,7 @@
 #include "MooseError.h"
 #include "Backup.h"
 #include "RankTwoTensor.h"
+#include "RankThreeTensor.h"
 #include "RankFourTensor.h"
 #include "MooseADWrapper.h"
 
@@ -398,6 +399,13 @@ dataStore(std::ostream & stream, RankTwoTensorTempl<T> & rtt, void * context)
 
 template <typename T>
 void
+dataStore(std::ostream & stream, RankThreeTensorTempl<T> & rtt, void * context)
+{
+  dataStore(stream, rtt._vals, context);
+}
+
+template <typename T>
+void
 dataStore(std::ostream & stream, RankFourTensorTempl<T> & rft, void * context)
 {
   dataStore(stream, rft._vals, context);
@@ -634,6 +642,13 @@ void
 dataLoad(std::istream & stream, RankTwoTensorTempl<T> & rtt, void * context)
 {
   dataLoad(stream, rtt._coords, context);
+}
+
+template <typename T>
+void
+dataLoad(std::istream & stream, RankThreeTensorTempl<T> & rtt, void * context)
+{
+  dataLoad(stream, rtt._vals, context);
 }
 
 template <typename T>

--- a/framework/include/utils/MooseADWrapper.h
+++ b/framework/include/utils/MooseADWrapper.h
@@ -4,6 +4,7 @@
 #include "MooseError.h"
 #include "DualReal.h"
 #include "RankTwoTensor.h"
+#include "RankThreeTensor.h"
 #include "RankFourTensor.h"
 
 #include "libmesh/vector_value.h"
@@ -234,6 +235,44 @@ private:
   friend void dataLoad<RankTwoTensorTempl<Real>>(std::istream &,
                                                  MooseADWrapper<RankTwoTensorTempl<Real>> &,
                                                  void *);
+};
+
+template <>
+class MooseADWrapper<RankThreeTensorTempl<Real>>
+{
+public:
+  MooseADWrapper(bool use_ad = false);
+  MooseADWrapper(MooseADWrapper<RankThreeTensorTempl<Real>> &&) = default;
+
+  typedef RankThreeTensorTempl<DualReal> DNType;
+
+  const RankThreeTensorTempl<Real> & value() const { return _val; }
+
+  RankThreeTensorTempl<Real> & value() { return _val; }
+
+  const RankThreeTensorTempl<DualReal> & dn(bool = true) const;
+
+  RankThreeTensorTempl<DualReal> & dn(bool = true);
+
+  void copyDualNumberToValue();
+
+  void markAD(bool use_ad);
+
+  MooseADWrapper<RankThreeTensorTempl<Real>> &
+  operator=(const MooseADWrapper<RankThreeTensorTempl<Real>> &);
+  MooseADWrapper<RankThreeTensorTempl<Real>> &
+  operator=(MooseADWrapper<RankThreeTensorTempl<Real>> &&) = default;
+
+private:
+  bool _use_ad;
+  RankThreeTensorTempl<Real> _val;
+  mutable std::unique_ptr<RankThreeTensorTempl<DualReal>> _dual_number;
+  friend void dataStore<RankThreeTensorTempl<Real>>(std::ostream &,
+                                                    MooseADWrapper<RankThreeTensorTempl<Real>> &,
+                                                    void *);
+  friend void dataLoad<RankThreeTensorTempl<Real>>(std::istream &,
+                                                   MooseADWrapper<RankThreeTensorTempl<Real>> &,
+                                                   void *);
 };
 
 template <>

--- a/framework/include/utils/RankFourTensor.h
+++ b/framework/include/utils/RankFourTensor.h
@@ -34,10 +34,8 @@ class VectorValue;
 class MooseEnum;
 template <typename>
 class RankTwoTensorTempl;
-typedef RankTwoTensorTempl<Real> RankTwoTensor;
 template <typename>
 class RankFourTensorTempl;
-typedef RankFourTensorTempl<Real> RankFourTensor;
 
 template <typename T>
 void mooseSetToZero(T & v);
@@ -412,8 +410,12 @@ protected:
   friend class RankTwoTensorTempl;
   template <typename T2>
   friend class RankFourTensorTempl;
-  friend class RankThreeTensor;
+  template <typename T2>
+  friend class RankThreeTensorTempl;
 };
+
+typedef RankFourTensorTempl<Real> RankFourTensor;
+typedef RankFourTensorTempl<DualReal> DualRankFourTensor;
 
 template <typename T1, typename T2>
 inline auto operator*(const T1 & a, const RankFourTensorTempl<T2> & b) ->

--- a/framework/include/utils/RankThreeTensor.h
+++ b/framework/include/utils/RankThreeTensor.h
@@ -76,6 +76,12 @@ public:
   /// Default constructor; fills to zero
   RankThreeTensorTempl();
 
+  /**
+   * Construct from other class template instantiation
+   */
+  template <typename T2>
+  RankThreeTensorTempl(const RankThreeTensorTempl<T2> & copy);
+
   /// Select specific initialization pattern
   RankThreeTensorTempl(const InitMethod);
 
@@ -93,6 +99,15 @@ public:
   {
     return _vals[((i * N + j) * N + k)];
   }
+
+  /**
+   * Assignment-from-scalar operator.  Used only to zero out the tensor.
+   *
+   * \returns A reference to *this.
+   */
+  template <typename Scalar>
+  typename boostcopy::enable_if_c<ScalarTraits<Scalar>::value, RankThreeTensorTempl &>::type
+  operator=(const Scalar & libmesh_dbg_var(p));
 
   /// Zeros out the tensor.
   void zero();
@@ -201,6 +216,9 @@ protected:
   friend class RankTwoTensorTempl;
 
   template <class T2>
+  friend class RankThreeTensorTempl;
+
+  template <class T2>
   friend class RankFourTensorTempl;
 };
 
@@ -208,23 +226,21 @@ typedef RankThreeTensorTempl<Real> RankThreeTensor;
 typedef RankThreeTensorTempl<DualReal> DualRankThreeTensor;
 
 template <typename T>
-RankThreeTensorTempl<T> operator*(T a, const RankThreeTensorTempl<T> & b)
+template <typename T2>
+RankThreeTensorTempl<T>::RankThreeTensorTempl(const RankThreeTensorTempl<T2> & copy)
 {
-  return b * a;
+  for (unsigned int i = 0; i < N3; ++i)
+    _vals[i] = copy._vals[i];
 }
 
-///r=v*A where r is rank 2, v is vector and A is rank 3
 template <typename T>
-RankTwoTensorTempl<T> operator*(const VectorValue<T> & p, const RankThreeTensorTempl<T> & b)
+template <typename Scalar>
+typename boostcopy::enable_if_c<ScalarTraits<Scalar>::value, RankThreeTensorTempl<T> &>::type
+RankThreeTensorTempl<T>::operator=(const Scalar & libmesh_dbg_var(p))
 {
-  RankTwoTensorTempl<T> result;
-
-  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
-    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
-      for (unsigned int k = 0; k < LIBMESH_DIM; ++k)
-        result(i, j) += p(k) * b(k, i, j);
-
-  return result;
+  libmesh_assert_equal_to(p, Scalar(0));
+  this->zero();
+  return *this;
 }
 
 template <typename T>
@@ -251,6 +267,26 @@ RankThreeTensorTempl<T>::rotate(const T2 & R)
         }
         _vals[index++] = sum;
       }
+}
+
+template <typename T>
+RankThreeTensorTempl<T> operator*(T a, const RankThreeTensorTempl<T> & b)
+{
+  return b * a;
+}
+
+///r=v*A where r is rank 2, v is vector and A is rank 3
+template <typename T>
+RankTwoTensorTempl<T> operator*(const VectorValue<T> & p, const RankThreeTensorTempl<T> & b)
+{
+  RankTwoTensorTempl<T> result;
+
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+      for (unsigned int k = 0; k < LIBMESH_DIM; ++k)
+        result(i, j) += p(k) * b(k, i, j);
+
+  return result;
 }
 
 #endif // RANKTHREETENSOR_H

--- a/framework/include/utils/RankThreeTensor.h
+++ b/framework/include/utils/RankThreeTensor.h
@@ -10,22 +10,32 @@
 #ifndef RANKTHREETENSOR_H
 #define RANKTHREETENSOR_H
 
-// MOOSE includes
-#include "DataIO.h"
+#include "Moose.h"
+#include "DualReal.h"
 
-#include "libmesh/tensor_value.h"
 #include "libmesh/libmesh.h"
-#include "libmesh/vector_value.h"
+
+using libMesh::Real;
+namespace libMesh
+{
+template <typename>
+class TensorValue;
+template <typename>
+class TypeTensor;
+template <typename>
+class VectorValue;
+}
 
 // Forward declarations
 class MooseEnum;
 template <typename>
+class RankThreeTensorTempl;
+
+template <typename>
 class RankTwoTensorTempl;
-typedef RankTwoTensorTempl<Real> RankTwoTensor;
-class RankThreeTensor;
+
 template <typename>
 class RankFourTensorTempl;
-typedef RankFourTensorTempl<Real> RankFourTensor;
 
 template <typename T>
 void mooseSetToZero(T & v);
@@ -35,13 +45,16 @@ void mooseSetToZero(T & v);
  * Needed by DerivativeMaterialInterface
  */
 template <>
-void mooseSetToZero<RankThreeTensor>(RankThreeTensor & v);
+void mooseSetToZero<RankThreeTensorTempl<Real>>(RankThreeTensorTempl<Real> & v);
+template <>
+void mooseSetToZero<RankThreeTensorTempl<DualReal>>(RankThreeTensorTempl<DualReal> & v);
 
 /**
  * RankThreeTensor is designed to handle any N-dimensional third order tensor, r.
  *
  */
-class RankThreeTensor
+template <typename T>
+class RankThreeTensorTempl
 {
 public:
   /// Initialization method
@@ -61,22 +74,22 @@ public:
   };
 
   /// Default constructor; fills to zero
-  RankThreeTensor();
+  RankThreeTensorTempl();
 
   /// Select specific initialization pattern
-  RankThreeTensor(const InitMethod);
+  RankThreeTensorTempl(const InitMethod);
 
   /// Fill from vector
-  RankThreeTensor(const std::vector<Real> &, FillMethod);
+  RankThreeTensorTempl(const std::vector<T> &, FillMethod);
 
   /// Gets the value for the index specified.  Takes index = 0,1,2
-  inline Real & operator()(unsigned int i, unsigned int j, unsigned int k)
+  inline T & operator()(unsigned int i, unsigned int j, unsigned int k)
   {
     return _vals[((i * N + j) * N + k)];
   }
 
   /// Gets the value for the index specified.  Takes index = 0,1,2. Used for const
-  inline Real operator()(unsigned int i, unsigned int j, unsigned int k) const
+  inline T operator()(unsigned int i, unsigned int j, unsigned int k) const
   {
     return _vals[((i * N + j) * N + k)];
   }
@@ -88,56 +101,53 @@ public:
   void print(std::ostream & stm = Moose::out) const;
 
   /// copies values from a into this tensor
-  RankThreeTensor & operator=(const RankThreeTensor & a);
-
-  /// r_ijk*a_kl
-  // RankTwoTensor operator*(const RankTwoTensor & a) const;
+  RankThreeTensorTempl<T> & operator=(const RankThreeTensorTempl<T> & a);
 
   /// b_i = r_ijk * a_jk
-  RealVectorValue operator*(const RankTwoTensor & a) const;
+  VectorValue<T> operator*(const RankTwoTensorTempl<T> & a) const;
 
   /// r_ijk*a
-  RankThreeTensor operator*(const Real a) const;
+  RankThreeTensorTempl<T> operator*(const T a) const;
 
   /// r_ijk *= a
-  RankThreeTensor & operator*=(const Real a);
+  RankThreeTensorTempl<T> & operator*=(const T a);
 
   /// r_ijk/a
-  RankThreeTensor operator/(const Real a) const;
+  RankThreeTensorTempl<T> operator/(const T a) const;
 
   /// r_ijk /= a  for all i, j, k
-  RankThreeTensor & operator/=(const Real a);
+  RankThreeTensorTempl<T> & operator/=(const T a);
 
   /// r_ijk += a_ijk  for all i, j, k
-  RankThreeTensor & operator+=(const RankThreeTensor & a);
+  RankThreeTensorTempl<T> & operator+=(const RankThreeTensorTempl<T> & a);
 
   /// r_ijkl + a_ijk
-  RankThreeTensor operator+(const RankThreeTensor & a) const;
+  RankThreeTensorTempl<T> operator+(const RankThreeTensorTempl<T> & a) const;
 
   /// r_ijk -= a_ijk
-  RankThreeTensor & operator-=(const RankThreeTensor & a);
+  RankThreeTensorTempl<T> & operator-=(const RankThreeTensorTempl<T> & a);
 
   /// r_ijk - a_ijk
-  RankThreeTensor operator-(const RankThreeTensor & a) const;
+  RankThreeTensorTempl<T> operator-(const RankThreeTensorTempl<T> & a) const;
 
   /// -r_ijk
-  RankThreeTensor operator-() const;
+  RankThreeTensorTempl<T> operator-() const;
 
   /// \sqrt(r_ijk*r_ijk)
-  Real L2norm() const;
+  T L2norm() const;
 
   /**
    * Rotate the tensor using
    * r_ijk = R_im R_in R_ko r_mno
    */
-  template <class T>
-  void rotate(const T & R);
+  template <class T2>
+  void rotate(const T2 & R);
 
   /**
    * Rotate the tensor using
    * r_ijk = R_im R_in R_ko r_mno
    */
-  void rotate(const RealTensorValue & R);
+  void rotate(const TensorValue<T> & R);
 
   /// Static method for use in validParams for getting the "fill_method"
   static MooseEnum fillMethodEnum();
@@ -150,25 +160,25 @@ public:
    *             general (use fillGeneralFromInputVector)
    *             more fill_methods to be implemented soon!
    */
-  void fillFromInputVector(const std::vector<Real> & input, FillMethod fill_method);
+  void fillFromInputVector(const std::vector<T> & input, FillMethod fill_method);
 
   /**
    * Fills RankThreeTensor from plane normal vectors
    * ref. Kuhl et. al. Int. J. Solids Struct. 38(2001) 2933-2952
    * @param input plane normal vector
    */
-  void fillFromPlaneNormal(const RealVectorValue & input);
+  void fillFromPlaneNormal(const VectorValue<T> & input);
 
   /**
    * Creates fourth order tensor D=A*b*A where A is rank 3 and b is rank 2
    * @param a RankTwoTensor A in the equation above
    */
-  RankFourTensor mixedProductRankFour(const RankTwoTensor & a) const;
+  RankFourTensorTempl<T> mixedProductRankFour(const RankTwoTensorTempl<T> & a) const;
 
   /**
    * Creates a vector from the double contraction of a rank three and rank two tensor.
    */
-  RealVectorValue doubleContraction(const RankTwoTensor & b) const;
+  VectorValue<T> doubleContraction(const RankTwoTensorTempl<T> & b) const;
 
 protected:
   /// Dimensionality of rank-three tensor
@@ -177,31 +187,40 @@ protected:
   static constexpr unsigned int N3 = N * N * N;
 
   /// The values of the rank-three tensor stored by index=((i * LIBMESH_DIM + j) * LIBMESH_DIM + k)
-  Real _vals[N3];
+  T _vals[N3];
 
-  void fillGeneralFromInputVector(const std::vector<Real> & input);
+  void fillGeneralFromInputVector(const std::vector<T> & input);
 
-  template <class T>
-  friend void dataStore(std::ostream &, T &, void *);
+  template <class T2>
+  friend void dataStore(std::ostream &, RankThreeTensorTempl<T2> &, void *);
 
-  template <class T>
-  friend void dataLoad(std::istream &, T &, void *);
+  template <class T2>
+  friend void dataLoad(std::istream &, RankThreeTensorTempl<T2> &, void *);
+
+  template <class T2>
+  friend class RankTwoTensorTempl;
+
+  template <class T2>
+  friend class RankFourTensorTempl;
 };
 
-template <>
-void dataStore(std::ostream &, RankThreeTensor &, void *);
+typedef RankThreeTensorTempl<Real> RankThreeTensor;
+typedef RankThreeTensorTempl<DualReal> DualRankThreeTensor;
 
-template <>
-void dataLoad(std::istream &, RankThreeTensor &, void *);
-
-inline RankThreeTensor operator*(Real a, const RankThreeTensor & b) { return b * a; }
+template <typename T>
+RankThreeTensorTempl<T> operator*(T a, const RankThreeTensorTempl<T> & b)
+{
+  return b * a;
+}
 
 ///r=v*A where r is rank 2, v is vector and A is rank 3
-RankTwoTensor operator*(const RealVectorValue &, const RankThreeTensor &);
+template <typename T>
+RankTwoTensorTempl<T> operator*(const VectorValue<T> &, const RankThreeTensorTempl<T> &);
 
-template <class T>
+template <typename T>
+template <class T2>
 void
-RankThreeTensor::rotate(const T & R)
+RankThreeTensorTempl<T>::rotate(const T2 & R)
 {
   unsigned int index = 0;
   for (unsigned int i = 0; i < N; ++i)
@@ -209,13 +228,13 @@ RankThreeTensor::rotate(const T & R)
       for (unsigned int k = 0; k < N; ++k)
       {
         unsigned int index2 = 0;
-        Real sum = 0.0;
+        T sum = 0.0;
         for (unsigned int m = 0; m < N; ++m)
         {
-          Real a = R(i, m);
+          T a = R(i, m);
           for (unsigned int n = 0; n < N; ++n)
           {
-            Real ab = a * R(j, n);
+            T ab = a * R(j, n);
             for (unsigned int o = 0; o < N; ++o)
               sum += ab * R(k, o) * _vals[index2++];
           }

--- a/framework/include/utils/RankTwoTensor.h
+++ b/framework/include/utils/RankTwoTensor.h
@@ -482,7 +482,8 @@ private:
   friend void dataLoad(std::istream &, RankTwoTensorTempl<T2> &, void *);
   template <class T2>
   friend class RankFourTensorTempl;
-  friend class RankThreeTensor;
+  template <class T2>
+  friend class RankThreeTensorTempl;
 };
 
 typedef RankTwoTensorTempl<Real> RankTwoTensor;

--- a/framework/src/restart/DataIO.C
+++ b/framework/src/restart/DataIO.C
@@ -58,6 +58,13 @@ dataStore(std::ostream & stream, RankTwoTensor & rtt, void * context)
 
 template <>
 void
+dataStore(std::ostream & stream, RankThreeTensor & rtht, void * context)
+{
+  dataStore(stream, rtht._vals, context);
+}
+
+template <>
+void
 dataStore(std::ostream & stream, RankFourTensor & rft, void * context)
 {
   dataStore(stream, rft._vals, context);

--- a/framework/src/utils/MooseADWrapper.C
+++ b/framework/src/utils/MooseADWrapper.C
@@ -210,6 +210,65 @@ operator=(const MooseADWrapper<RankTwoTensorTempl<Real>> & rhs)
   return *this;
 }
 
+// RankThreeTensor
+MooseADWrapper<RankThreeTensorTempl<Real>>::MooseADWrapper(bool use_ad)
+  : _use_ad(use_ad), _val(), _dual_number(nullptr)
+{
+  if (_use_ad)
+    _dual_number = libmesh_make_unique<RankThreeTensorTempl<DualReal>>();
+}
+
+const RankThreeTensorTempl<DualReal> &
+MooseADWrapper<RankThreeTensorTempl<Real>>::dn(bool) const
+{
+  if (!_dual_number)
+    _dual_number = libmesh_make_unique<RankThreeTensorTempl<DualReal>>(_val);
+  else if (!_use_ad)
+    for (std::size_t i = 0; i < LIBMESH_DIM; ++i)
+      for (std::size_t j = 0; j < LIBMESH_DIM; ++j)
+        for (std::size_t k = 0; k < LIBMESH_DIM; ++k)
+          (*_dual_number)(i, j, k).value() = _val(i, j, k);
+  return *_dual_number;
+}
+
+RankThreeTensorTempl<DualReal> &
+MooseADWrapper<RankThreeTensorTempl<Real>>::dn(bool)
+{
+  return *_dual_number;
+}
+
+void
+MooseADWrapper<RankThreeTensorTempl<Real>>::copyDualNumberToValue()
+{
+  for (std::size_t i = 0; i < LIBMESH_DIM; ++i)
+    for (std::size_t j = 0; j < LIBMESH_DIM; ++j)
+      for (std::size_t k = 0; k < LIBMESH_DIM; ++k)
+        _val(i, j, k) = (*_dual_number)(i, j, k).value();
+}
+
+void
+MooseADWrapper<RankThreeTensorTempl<Real>>::markAD(bool use_ad)
+{
+  if (!use_ad && _use_ad)
+    _dual_number = nullptr;
+  else if (use_ad && !_use_ad)
+    _dual_number = libmesh_make_unique<RankThreeTensorTempl<DualReal>>();
+  _use_ad = use_ad;
+}
+
+MooseADWrapper<RankThreeTensorTempl<Real>> &
+MooseADWrapper<RankThreeTensorTempl<Real>>::
+operator=(const MooseADWrapper<RankThreeTensorTempl<Real>> & rhs)
+{
+  _val = rhs._val;
+  if (_dual_number && rhs._dual_number)
+    *_dual_number = *rhs._dual_number;
+  else if (_dual_number)
+    *_dual_number = 0;
+  return *this;
+}
+
+// RankFourTensor
 MooseADWrapper<RankFourTensorTempl<Real>>::MooseADWrapper(bool use_ad)
   : _use_ad(use_ad), _val(), _dual_number(nullptr)
 {

--- a/framework/src/utils/RankThreeTensor.C
+++ b/framework/src/utils/RankThreeTensor.C
@@ -19,6 +19,11 @@
 #include "PermutationTensor.h"
 
 #include "libmesh/utility.h"
+#include "libmesh/vector_value.h"
+#include "libmesh/tensor_value.h"
+
+#include "metaphysicl/numberarray.h"
+#include "metaphysicl/dualnumber.h"
 
 // C++ includes
 #include <iomanip>
@@ -26,32 +31,27 @@
 
 template <>
 void
-mooseSetToZero<RankThreeTensor>(RankThreeTensor & v)
+mooseSetToZero<RankThreeTensorTempl<Real>>(RankThreeTensorTempl<Real> & v)
 {
   v.zero();
 }
 
 template <>
 void
-dataStore(std::ostream & stream, RankThreeTensor & rtht, void * context)
+mooseSetToZero<RankThreeTensorTempl<DualReal>>(RankThreeTensorTempl<DualReal> & v)
 {
-  dataStore(stream, rtht._vals, context);
+  v.zero();
 }
 
-template <>
-void
-dataLoad(std::istream & stream, RankThreeTensor & rtht, void * context)
-{
-  dataLoad(stream, rtht._vals, context);
-}
-
-MooseEnum RankThreeTensor::fillMethodEnum() // TODO: Need new fillMethodEnum() -- for now we will
-                                            // just use general (at most 27 components)
+template <typename T>
+MooseEnum RankThreeTensorTempl<T>::fillMethodEnum() // TODO: Need new fillMethodEnum() -- for now we
+                                                    // will just use general (at most 27 components)
 {
   return MooseEnum("general");
 }
 
-RankThreeTensor::RankThreeTensor()
+template <typename T>
+RankThreeTensorTempl<T>::RankThreeTensorTempl()
 {
   mooseAssert(N == 3, "RankThreeTensor is currently only tested for 3 dimensions.");
 
@@ -59,7 +59,8 @@ RankThreeTensor::RankThreeTensor()
     _vals[i] = 0;
 }
 
-RankThreeTensor::RankThreeTensor(const InitMethod init)
+template <typename T>
+RankThreeTensorTempl<T>::RankThreeTensorTempl(const InitMethod init)
 {
   switch (init)
   {
@@ -71,20 +72,23 @@ RankThreeTensor::RankThreeTensor(const InitMethod init)
   }
 }
 
-RankThreeTensor::RankThreeTensor(const std::vector<Real> & input, FillMethod fill_method)
+template <typename T>
+RankThreeTensorTempl<T>::RankThreeTensorTempl(const std::vector<T> & input, FillMethod fill_method)
 {
   fillFromInputVector(input, fill_method);
 }
 
+template <typename T>
 void
-RankThreeTensor::zero()
+RankThreeTensorTempl<T>::zero()
 {
   for (unsigned int i = 0; i < N3; ++i)
     _vals[i] = 0;
 }
 
-RankThreeTensor &
-RankThreeTensor::operator=(const RankThreeTensor & a)
+template <typename T>
+RankThreeTensorTempl<T> &
+RankThreeTensorTempl<T>::operator=(const RankThreeTensorTempl<T> & a)
 {
   for (unsigned int i = 0; i < N3; ++i)
     _vals[i] = a._vals[i];
@@ -92,13 +96,14 @@ RankThreeTensor::operator=(const RankThreeTensor & a)
   return *this;
 }
 
-RealVectorValue RankThreeTensor::operator*(const RankTwoTensor & a) const
+template <typename T>
+VectorValue<T> RankThreeTensorTempl<T>::operator*(const RankTwoTensorTempl<T> & a) const
 {
-  RealVectorValue result;
+  VectorValue<T> result;
 
   for (unsigned int i = 0; i < N; ++i)
   {
-    Real sum = 0;
+    T sum = 0;
     unsigned int i1 = i * N2;
     for (unsigned int j1 = 0; j1 < N2; j1 += N)
       for (unsigned int k = 0; k < N; ++k)
@@ -109,9 +114,10 @@ RealVectorValue RankThreeTensor::operator*(const RankTwoTensor & a) const
   return result;
 }
 
-RankThreeTensor RankThreeTensor::operator*(const Real b) const
+template <typename T>
+RankThreeTensorTempl<T> RankThreeTensorTempl<T>::operator*(const T b) const
 {
-  RankThreeTensor result;
+  RankThreeTensorTempl<T> result;
 
   for (unsigned int i = 0; i < N3; ++i)
     result._vals[i] = _vals[i] * b;
@@ -119,8 +125,9 @@ RankThreeTensor RankThreeTensor::operator*(const Real b) const
   return result;
 }
 
-RankThreeTensor &
-RankThreeTensor::operator*=(const Real a)
+template <typename T>
+RankThreeTensorTempl<T> &
+RankThreeTensorTempl<T>::operator*=(const T a)
 {
   for (unsigned int i = 0; i < N3; ++i)
     _vals[i] *= a;
@@ -128,10 +135,11 @@ RankThreeTensor::operator*=(const Real a)
   return *this;
 }
 
-RankThreeTensor
-RankThreeTensor::operator/(const Real b) const
+template <typename T>
+RankThreeTensorTempl<T>
+RankThreeTensorTempl<T>::operator/(const T b) const
 {
-  RankThreeTensor result;
+  RankThreeTensorTempl<T> result;
 
   for (unsigned int i = 0; i < N3; ++i)
     result._vals[i] = _vals[i] / b;
@@ -139,8 +147,9 @@ RankThreeTensor::operator/(const Real b) const
   return result;
 }
 
-RankThreeTensor &
-RankThreeTensor::operator/=(const Real a)
+template <typename T>
+RankThreeTensorTempl<T> &
+RankThreeTensorTempl<T>::operator/=(const T a)
 {
   for (unsigned int i = 0; i < N3; ++i)
     _vals[i] /= a;
@@ -148,8 +157,9 @@ RankThreeTensor::operator/=(const Real a)
   return *this;
 }
 
-RankThreeTensor &
-RankThreeTensor::operator+=(const RankThreeTensor & a)
+template <typename T>
+RankThreeTensorTempl<T> &
+RankThreeTensorTempl<T>::operator+=(const RankThreeTensorTempl<T> & a)
 {
   for (unsigned int i = 0; i < N3; ++i)
     _vals[i] += a._vals[i];
@@ -157,10 +167,11 @@ RankThreeTensor::operator+=(const RankThreeTensor & a)
   return *this;
 }
 
-RankThreeTensor
-RankThreeTensor::operator+(const RankThreeTensor & b) const
+template <typename T>
+RankThreeTensorTempl<T>
+RankThreeTensorTempl<T>::operator+(const RankThreeTensorTempl<T> & b) const
 {
-  RankThreeTensor result;
+  RankThreeTensorTempl<T> result;
 
   for (unsigned int i = 0; i < N3; ++i)
     result._vals[i] = _vals[i] + b._vals[i];
@@ -168,8 +179,9 @@ RankThreeTensor::operator+(const RankThreeTensor & b) const
   return result;
 }
 
-RankThreeTensor &
-RankThreeTensor::operator-=(const RankThreeTensor & a)
+template <typename T>
+RankThreeTensorTempl<T> &
+RankThreeTensorTempl<T>::operator-=(const RankThreeTensorTempl<T> & a)
 {
   for (unsigned int i = 0; i < N3; ++i)
     _vals[i] -= a._vals[i];
@@ -177,10 +189,11 @@ RankThreeTensor::operator-=(const RankThreeTensor & a)
   return *this;
 }
 
-RankThreeTensor
-RankThreeTensor::operator-(const RankThreeTensor & b) const
+template <typename T>
+RankThreeTensorTempl<T>
+RankThreeTensorTempl<T>::operator-(const RankThreeTensorTempl<T> & b) const
 {
-  RankThreeTensor result;
+  RankThreeTensorTempl<T> result;
 
   for (unsigned int i = 0; i < N3; ++i)
     result._vals[i] = _vals[i] - b._vals[i];
@@ -188,10 +201,11 @@ RankThreeTensor::operator-(const RankThreeTensor & b) const
   return result;
 }
 
-RankThreeTensor
-RankThreeTensor::operator-() const
+template <typename T>
+RankThreeTensorTempl<T>
+RankThreeTensorTempl<T>::operator-() const
 {
-  RankThreeTensor result;
+  RankThreeTensorTempl<T> result;
 
   for (unsigned int i = 0; i < N3; ++i)
     result._vals[i] = -_vals[i];
@@ -199,10 +213,11 @@ RankThreeTensor::operator-() const
   return result;
 }
 
-Real
-RankThreeTensor::L2norm() const
+template <typename T>
+T
+RankThreeTensorTempl<T>::L2norm() const
 {
-  Real l2 = 0;
+  T l2 = 0;
 
   for (unsigned int i = 0; i < N3; ++i)
     l2 += Utility::pow<2>(_vals[i]);
@@ -210,8 +225,9 @@ RankThreeTensor::L2norm() const
   return std::sqrt(l2);
 }
 
+template <typename T>
 void
-RankThreeTensor::fillFromInputVector(const std::vector<Real> & input, FillMethod fill_method)
+RankThreeTensorTempl<T>::fillFromInputVector(const std::vector<T> & input, FillMethod fill_method)
 {
   zero();
 
@@ -225,20 +241,21 @@ RankThreeTensor::fillFromInputVector(const std::vector<Real> & input, FillMethod
   }
 }
 
+template <typename T>
 void
-RankThreeTensor::fillFromPlaneNormal(const RealVectorValue & input)
+RankThreeTensorTempl<T>::fillFromPlaneNormal(const VectorValue<T> & input)
 {
   unsigned int index = 0;
   for (unsigned int i = 0; i < N; ++i)
   {
-    const Real a = input(i);
+    const T a = input(i);
     for (unsigned int j = 0; j < N; ++j)
     {
-      const Real b = input(j);
+      const T b = input(j);
       for (unsigned int k = 0; k < N; ++k)
       {
-        const Real c = input(k);
-        Real sum = 0;
+        const T c = input(k);
+        T sum = 0;
         sum = -2 * a * b * c;
         if (i == j)
           sum += c;
@@ -250,10 +267,11 @@ RankThreeTensor::fillFromPlaneNormal(const RealVectorValue & input)
   }
 }
 
-RankFourTensor
-RankThreeTensor::mixedProductRankFour(const RankTwoTensor & a) const
+template <typename T>
+RankFourTensorTempl<T>
+RankThreeTensorTempl<T>::mixedProductRankFour(const RankTwoTensorTempl<T> & a) const
 {
-  RankFourTensor result;
+  RankFourTensorTempl<T> result;
 
   unsigned int index = 0;
   for (unsigned int i = 0; i < N; ++i)
@@ -270,24 +288,25 @@ RankThreeTensor::mixedProductRankFour(const RankTwoTensor & a) const
   return result;
 }
 
+template <typename T>
 void
-RankThreeTensor::rotate(const RealTensorValue & R)
+RankThreeTensorTempl<T>::rotate(const TensorValue<T> & R)
 {
-  RankThreeTensor old = *this;
+  RankThreeTensorTempl<T> old = *this;
 
   unsigned int index = 0;
   for (unsigned int i = 0; i < N; ++i)
     for (unsigned int j = 0; j < N; ++j)
       for (unsigned int k = 0; k < N; ++k)
       {
-        Real sum = 0.0;
+        T sum = 0.0;
         unsigned int index2 = 0;
         for (unsigned int m = 0; m < N; ++m)
         {
-          Real a = R(i, m);
+          T a = R(i, m);
           for (unsigned int n = 0; n < N; ++n)
           {
-            Real ab = a * R(j, n);
+            T ab = a * R(j, n);
             for (unsigned int o = 0; o < N; ++o)
               sum += ab * R(k, o) * old._vals[index2++];
           }
@@ -296,8 +315,9 @@ RankThreeTensor::rotate(const RealTensorValue & R)
       }
 }
 
+template <typename T>
 void
-RankThreeTensor::fillGeneralFromInputVector(const std::vector<Real> & input)
+RankThreeTensorTempl<T>::fillGeneralFromInputVector(const std::vector<T> & input)
 {
   if (input.size() != 27)
     mooseError("To use fillGeneralFromInputVector, your input must have size 27. Yours has size ",
@@ -307,9 +327,10 @@ RankThreeTensor::fillGeneralFromInputVector(const std::vector<Real> & input)
     _vals[i] = input[i];
 }
 
-RankTwoTensor operator*(const RealVectorValue & p, const RankThreeTensor & b)
+template <typename T>
+RankTwoTensorTempl<T> operator*(const VectorValue<T> & p, const RankThreeTensorTempl<T> & b)
 {
-  RankTwoTensor result;
+  RankTwoTensorTempl<T> result;
 
   for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
     for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
@@ -319,10 +340,11 @@ RankTwoTensor operator*(const RealVectorValue & p, const RankThreeTensor & b)
   return result;
 }
 
-RealVectorValue
-RankThreeTensor::doubleContraction(const RankTwoTensor & b) const
+template <typename T>
+VectorValue<T>
+RankThreeTensorTempl<T>::doubleContraction(const RankTwoTensorTempl<T> & b) const
 {
-  RealVectorValue result;
+  VectorValue<T> result;
 
   for (unsigned int i = 0; i < N; ++i)
     for (unsigned int j = 0; j < N2; ++j)
@@ -330,3 +352,6 @@ RankThreeTensor::doubleContraction(const RankTwoTensor & b) const
 
   return result;
 }
+
+template class RankThreeTensorTempl<Real>;
+template class RankThreeTensorTempl<DualReal>;

--- a/modules/phase_field/include/kernels/ACInterfaceStress.h
+++ b/modules/phase_field/include/kernels/ACInterfaceStress.h
@@ -11,12 +11,15 @@
 #define ACINTERFACESTRESS_H
 
 #include "Kernel.h"
-#include "RankThreeTensor.h"
 
 class ACInterfaceStress;
 template <typename>
 class RankTwoTensorTempl;
 typedef RankTwoTensorTempl<Real> RankTwoTensor;
+
+template <typename>
+class RankThreeTensorTempl;
+typedef RankThreeTensorTempl<Real> RankThreeTensor;
 
 template <>
 InputParameters validParams<ACInterfaceStress>();

--- a/modules/phase_field/src/kernels/ACInterfaceStress.C
+++ b/modules/phase_field/src/kernels/ACInterfaceStress.C
@@ -9,6 +9,7 @@
 
 #include "ACInterfaceStress.h"
 #include "RankTwoTensor.h"
+#include "RankThreeTensor.h"
 
 registerMooseObject("PhaseFieldApp", ACInterfaceStress);
 

